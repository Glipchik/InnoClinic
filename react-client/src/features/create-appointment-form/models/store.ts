import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AxiosResponse, AxiosError } from 'axios';
import { takeLatest } from 'redux-saga/effects';
import { call, CallEffect, put, PutEffect } from 'redux-saga/effects';
import TimeSlot from '../../../entities/timeSlot';
import { AnyAction } from 'redux-saga';
import CreateAppointmentModel from '../../../features/create-appointment-form/models/createAppointmentModel';
import appointmentsApi from '../../../shared/api/appointments';

interface CreateAppointmentState {
  loading: boolean
  error?: string
}

const initialState : CreateAppointmentState = {
  loading: false,
  error: undefined
};

const createAppointmentSlice = createSlice({
  name: 'CreateAppointmentSlice',
  initialState,
  reducers: {
    createAppointmentRequest: (state, action: PayloadAction<CreateAppointmentModel>) => {
      state.loading = true;
    },
    createAppointmentSuccess: (state) => {
      state.loading = false;
    },
    createAppointmentFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

type ApiResponse = AxiosResponse<TimeSlot[]>;

type ApiError = AxiosError<{
  detail?: string;
  title?: string;
}>;

function* createAppointment(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const createAppointment : CreateAppointmentModel = action.payload;
    const response = yield call(appointmentsApi.POST, createAppointment);
    yield put(createAppointmentSlice.actions.createAppointmentSuccess(response.data))
  } catch (error) {
    let errorMessage = "An unknown error occurred";

    if ((error as ApiError).response && (error as ApiError).response?.data) {
      const problemDetails = (error as ApiError).response!.data;
      errorMessage = problemDetails.detail || problemDetails.title || errorMessage;
    }
    yield put(createAppointmentSlice.actions.createAppointmentFailure(errorMessage))
  }
}

function* watchCreateAppointment() {
  yield takeLatest(createAppointmentSlice.actions.createAppointmentRequest.type, createAppointment);
}

export const { createAppointmentRequest } = createAppointmentSlice.actions;
export const createAppointmentSliceReducer = createAppointmentSlice.reducer
export { watchCreateAppointment }