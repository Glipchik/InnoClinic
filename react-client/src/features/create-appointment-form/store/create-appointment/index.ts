import appointmentsApi from '@features/create-appointment-form/api/appointments';
import CreateAppointmentModel from '@features/create-appointment-form/models/createAppointmentModel';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface CreateAppointmentState {
  success: boolean
  loading: boolean
  error?: string
}

const initialState : CreateAppointmentState = {
  success: false,
  loading: false,
  error: undefined
};

const createAppointmentSlice = createSlice({
  name: 'CreateAppointmentSlice',
  initialState,
  reducers: {
    createAppointmentRequest: (state, action: PayloadAction<CreateAppointmentModel>) => {
      state.loading = !!action;
    },
    createAppointmentSuccess: (state) => {
      state.loading = false;
      state.success = true
    },
    createAppointmentFailure: (state, action) => {
      state.success = false
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { createAppointmentRequest } = createAppointmentSlice.actions;
export const createAppointmentSliceReducer = createAppointmentSlice.reducer

type ApiResponse = AxiosResponse<CreateAppointmentModel>;

function* createAppointment(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const createAppointmentModel : CreateAppointmentModel = action.payload;
    yield call(appointmentsApi.post, createAppointmentModel);
    yield put(createAppointmentSlice.actions.createAppointmentSuccess())
  } catch (error) {
    yield put(createAppointmentSlice.actions.createAppointmentFailure(handleError(error as ApiError)))
  }
}

function* watchCreateAppointment() {
  yield takeLatest(createAppointmentRequest.type, createAppointment);
}

export { watchCreateAppointment }