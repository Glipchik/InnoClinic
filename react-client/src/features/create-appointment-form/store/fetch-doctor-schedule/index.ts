import TimeSlot from '@entities/timeSlot';
import timeSlotsApi from '@features/create-appointment-form/api/doctor-schedule';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError from '@shared/lib/errorHandler';
import { AxiosResponse, AxiosError } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface DoctorScheduleState {
  loading: boolean
  error?: string
  data?: TimeSlot[]
}

const initialState : DoctorScheduleState = {
  loading: false
};

const fetchDoctorScheduleSlice = createSlice({
  name: 'FetchDoctorScheduleSlice',
  initialState,
  reducers: {
    fetchDoctorScheduleRequest: (state, action: PayloadAction<{ doctorId: string, date: string }>) => {
      state.loading = !!action;
    },
    fetchDoctorScheduleSuccess: (state, action: PayloadAction<TimeSlot[]>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchDoctorScheduleFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { fetchDoctorScheduleRequest } = fetchDoctorScheduleSlice.actions;
export const fetchDoctorScheduleSliceReducer = fetchDoctorScheduleSlice.reducer

type ApiResponse = AxiosResponse<TimeSlot[]>;

type ApiError = AxiosError<{
  detail?: string;
  title?: string;
}>;

function* fetchDoctorSchedule(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const { doctorId, date } = action.payload;
    const response: ApiResponse = yield call(timeSlotsApi.get, doctorId, date);
    yield put(fetchDoctorScheduleSlice.actions.fetchDoctorScheduleSuccess(response.data))
  } catch (error) {
    yield put(fetchDoctorScheduleSlice.actions.fetchDoctorScheduleFailure(handleError(error as ApiError)))
  }
}

function* watchFetchDoctorSchedule() {
  yield takeLatest(fetchDoctorScheduleRequest.type, fetchDoctorSchedule);
}

export { watchFetchDoctorSchedule }