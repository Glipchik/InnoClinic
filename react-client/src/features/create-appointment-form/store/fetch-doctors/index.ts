import doctorsApi from '@features/create-appointment-form/api/doctors';
import DoctorModel from '@models/doctors/doctorModel';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { call, CallEffect, put, PutEffect, takeLatest } from 'redux-saga/effects';

interface DoctorsState {
  loading: boolean
  error?: string
  data?: DoctorModel[]
}

const initialState : DoctorsState = {
  loading: false,
  error: undefined,
  data: undefined
};

const fetchDoctorsSlice = createSlice({
  name: 'FetchDoctorsSlice',
  initialState,
  reducers: {
    fetchDoctorsRequest: (state, action: PayloadAction<{ specializationId: string }>) => {
      state.loading = !!action;
    },
    fetchDoctorsSuccess: (state, action: PayloadAction<DoctorModel[]>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchDoctorsFailure: (state, action: PayloadAction<string>) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { fetchDoctorsRequest } = fetchDoctorsSlice.actions;
export const fetchDoctorsSliceReducer = fetchDoctorsSlice.reducer

type ApiResponse = AxiosResponse<DoctorModel[]>;

function* fetchDoctors(action: AnyAction) : Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const {specializationId} = action.payload
    const response : ApiResponse  = yield call(doctorsApi.getAll, specializationId);
    yield put(fetchDoctorsSlice.actions.fetchDoctorsSuccess(response.data))
  } catch (error) {
    yield put(fetchDoctorsSlice.actions.fetchDoctorsFailure(handleError(error as ApiError)))
  }
}

function* watchFetchDoctors() {
  yield takeLatest(fetchDoctorsRequest.type, fetchDoctors);
}

export { watchFetchDoctors }