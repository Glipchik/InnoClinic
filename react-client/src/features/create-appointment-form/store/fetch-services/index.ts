import servicesApi from '@features/create-appointment-form/api/services';
import ServiceModel from '@models/services/serviceModel';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { call, CallEffect, put, PutEffect, takeLatest } from 'redux-saga/effects';

interface ServicesState {
  loading: boolean
  error?: string
  data?: ServiceModel[]
}

const initialState : ServicesState = {
  loading: false,
  error: undefined,
  data: undefined
};

const fetchServicesSlice = createSlice({
  name: 'FetchServicesSlice',
  initialState,
  reducers: {
    fetchServicesRequest: (state, action: PayloadAction<{ specializationId: string }>) => {
      state.loading = !!action;
    },
    fetchServicesSuccess: (state, action: PayloadAction<ServiceModel[]>) => {
      state.loading = false;
      state.data = action.payload;
    },
    fetchServicesFailure: (state, action) => {
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { fetchServicesRequest } = fetchServicesSlice.actions;
export const fetchServicesSliceReducer = fetchServicesSlice.reducer

type ApiResponse = AxiosResponse<ServiceModel[]>;

function* fetchServices(action: AnyAction) : Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const {specializationId} = action.payload
    const response : ApiResponse  = yield call(servicesApi.getAll, specializationId);
    yield put(fetchServicesSlice.actions.fetchServicesSuccess(response.data))
  } catch (error) {
    yield put(fetchServicesSlice.actions.fetchServicesFailure(handleError(error as ApiError)))
  }
}

function* watchFetchServices() {
  yield takeLatest(fetchServicesRequest.type, fetchServices);
}

export { watchFetchServices }