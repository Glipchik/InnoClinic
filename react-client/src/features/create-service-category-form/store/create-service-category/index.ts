import serviceCategoriesApi from '@features/create-service-category-form/api/service-categories';
import CreateServiceCategoryModel from '@features/create-service-category-form/models/createServiceCategoryModel';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface CreateServiceCategoryState {
  success: boolean
  loading: boolean
  error?: string
}

const initialState : CreateServiceCategoryState = {
  success: false,
  loading: false,
  error: undefined
};

const createServiceCategorySlice = createSlice({
  name: 'CreateServiceCategorySlice',
  initialState,
  reducers: {
    createServiceCategoryRequest: (state, action: PayloadAction<CreateServiceCategoryModel>) => {
      state.loading = !!action;
    },
    createServiceCategorySuccess: (state) => {
      state.loading = false;
      state.success = true
    },
    createServiceCategoryFailure: (state, action) => {
      state.success = false
      state.loading = false;
      state.error = action.payload;
    }
  }
});

export const { createServiceCategoryRequest } = createServiceCategorySlice.actions;
export const createServiceCategorySliceReducer = createServiceCategorySlice.reducer

type ApiResponse = AxiosResponse<CreateServiceCategoryModel>;

function* createServiceCategory(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const createServiceCategoryModel : CreateServiceCategoryModel = action.payload;
    yield call(serviceCategoriesApi.post, createServiceCategoryModel);
    yield put(createServiceCategorySlice.actions.createServiceCategorySuccess())
  } catch (error) {
    yield put(createServiceCategorySlice.actions.createServiceCategoryFailure(handleError(error as ApiError)))
  }
}

function* watchCreateServiceCategory() {
  yield takeLatest(createServiceCategoryRequest.type, createServiceCategory);
}

export { watchCreateServiceCategory }