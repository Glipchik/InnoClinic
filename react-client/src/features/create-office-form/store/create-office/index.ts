import officesApi from '@features/create-office-form/api/offices';
import { CreateOfficeModel } from '@features/create-office-form/models/createOfficeModel';
import { AnyAction, createSlice, PayloadAction } from '@reduxjs/toolkit';
import handleError, { ApiError } from '@shared/lib/errorHandler';
import { AxiosResponse } from 'axios';
import { CallEffect, PutEffect, call, put, takeLatest } from 'redux-saga/effects';

interface CreateOfficeState {
  success?: boolean
  loading: boolean
  error?: string
}

const initialState : CreateOfficeState = {
  success: false,
  loading: false,
  error: undefined
};

const createOfficeSlice = createSlice({
  name: 'CreateOfficeSlice',
  initialState,
  reducers: {
    createOfficeRequest: (state, action: PayloadAction<CreateOfficeModel>) => {
      state.loading = !!action;
    },
    createOfficeSuccess: (state) => {
      state.loading = false;
      state.success = true
    },
    createOfficeFailure: (state, action) => {
      state.success = false
      state.loading = false;
      state.error = action.payload;
    },
    resetState: (state) => {
      state.success = undefined
      state.loading = false;
      state.error = undefined;
    }
  }
});

export const { createOfficeRequest, resetState } = createOfficeSlice.actions;
export const createOfficeSliceReducer = createOfficeSlice.reducer

type ApiResponse = AxiosResponse<CreateOfficeModel>;

function* createOffice(
  action: AnyAction
): Generator<CallEffect<ApiResponse> | PutEffect, void, ApiResponse> {
  try {
    const createOfficeModel : CreateOfficeModel = action.payload;
    yield call(officesApi.post, createOfficeModel);
    yield put(createOfficeSlice.actions.createOfficeSuccess())
  } catch (error) {
    yield put(createOfficeSlice.actions.createOfficeFailure(handleError(error as ApiError)))
  }
}

function* watchCreateOffice() {
  yield takeLatest(createOfficeRequest.type, createOffice);
}

export { watchCreateOffice }