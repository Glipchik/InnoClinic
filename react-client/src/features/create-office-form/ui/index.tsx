import { Formik } from "formik";
import { useDispatch, useSelector } from "react-redux";
import { CreateOfficeModel } from "../models/createOfficeModel";
import { validationSchema } from "../models/validationSchema";
import { createOfficeRequest, resetState } from "../store/create-office";
import InnerForm from "./inner-form";
import { RootState } from "@app/store";
import { useState, useEffect } from "react";

interface CreateOfficeFormProps {
  close: () => void;
}

const CreateOfficeForm = ({ close }: CreateOfficeFormProps) => {
  const dispatch = useDispatch();
  const [isSubmited, setIsSubmited] = useState(false);
  const { success } = useSelector((state: RootState) => state.createOffice);

  useEffect(() => {
    dispatch(resetState());
  }, []);

  useEffect(() => {
    if (success && isSubmited) {
      close()
    }
  }, [success, isSubmited]);

  const onSubmit = (values: CreateOfficeModel) => {
    dispatch(createOfficeRequest(values));
    setIsSubmited(true);
  };

  const initialValues: CreateOfficeModel = {
    address: "",
    isActive: true,
    registryPhoneNumber: ""
  }

  return (
    <Formik initialValues={initialValues} validationSchema={validationSchema} onSubmit={onSubmit}>
      <InnerForm close={close} />
    </Formik>
  );
};

export { CreateOfficeForm };