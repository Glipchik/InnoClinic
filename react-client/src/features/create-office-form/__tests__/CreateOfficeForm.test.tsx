import { render, waitFor } from "@testing-library/react";
import { CreateOfficeForm } from "@features/create-office-form";
import { Provider } from "react-redux";
import { Store } from "@reduxjs/toolkit";
import { MemoryRouter } from "react-router-dom";
import { createMockStore, mockOnClose } from "./__mocks__";
import userEvent from "@testing-library/user-event";
import store from "@app/store";
import officesApi from "../api/offices";

const renderCreateOfficeForm = (store: Store) =>
  render(
    <Provider store={store}>
      <MemoryRouter>
        <CreateOfficeForm close={mockOnClose} />
      </MemoryRouter>
    </Provider>
  );

describe("CreateOfficeForm", () => {
  it("renders", async () => {
    const { getByTestId } = renderCreateOfficeForm(
      createMockStore({
        loading: false,
      })
    );

    await waitFor(() => {
      expect(getByTestId("create-office-form")).toBeInTheDocument();
      expect(getByTestId("create-office-form")).toBeInTheDocument();
      expect(
        getByTestId("create-office-form-address-input")
      ).toBeInTheDocument();
      expect(
        getByTestId("create-office-form-phone-number-input")
      ).toBeInTheDocument();
      expect(
        getByTestId("create-office-form-is-active-checkbox")
      ).toBeInTheDocument();
      expect(getByTestId("submit-button")).toBeInTheDocument();
      expect(getByTestId("cancel-button")).toBeInTheDocument();
      expect(
        getByTestId("create-office-form-is-active-checkbox")
      ).toBeInTheDocument();
    });
  });

  it("calls onClose when close button is clicked", async () => {
    const { getByTestId } = renderCreateOfficeForm(
      createMockStore({
        loading: false,
      })
    );

    await waitFor(async () => {
      const closeButton = getByTestId("cancel-button");
      await userEvent.click(closeButton);
      expect(mockOnClose).toHaveBeenCalledTimes(1);
    });
  });

  it("calls submit handler when form is submitted with valid data", async () => {
    const { getByTestId } = renderCreateOfficeForm(store);
    const postSpy = jest.spyOn(officesApi, "post");
    const addressInput = getByTestId(
      "create-office-form-address-input"
    ) as HTMLInputElement;
    const phoneInput = getByTestId(
      "create-office-form-phone-number-input"
    ) as HTMLInputElement;
    const submitButton = getByTestId("submit-button");

    await userEvent.type(addressInput, "123 Main St");
    await userEvent.type(phoneInput, "+375293061111");

    await userEvent.click(submitButton);

    await waitFor(() => {
      expect(addressInput.value).toBe("123 Main St");
      expect(phoneInput.value).toBe("+375293061111");
      expect(postSpy).toHaveBeenCalledWith({
        address: "123 Main St",
        registryPhoneNumber: "+375293061111",
        isActive: true,
      });
    });
  });

  it("displays validation errors when form is submitted with invalid data", async () => {
    const store = createMockStore({
      loading: false,
      success: false,
    });
    const { getByTestId, getByText } = renderCreateOfficeForm(store);

    const submitButton = getByTestId("submit-button");
    await userEvent.click(submitButton);

    await waitFor(() => {
      expect(getByText("Address is required")).toBeInTheDocument();
      expect(getByText("Phone number is required")).toBeInTheDocument();
    });
  });
});
