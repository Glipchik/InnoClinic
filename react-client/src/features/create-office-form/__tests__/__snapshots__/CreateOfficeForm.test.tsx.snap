// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOfficeForm matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-full flex justify-center"
    data-testid="create-office-form"
  >
    <form
      action="#"
      class="flex w-[40%] flex-col gap-6 p-6 bg-white shadow-lg rounded-lg max-w-lg m-6"
    >
      <div
        class="form-group flex flex-col gap-y-2"
      >
        <label
          class="text-gray-700 font-medium"
          for="input-field"
        >
          Address
        </label>
        <input
          class="form-control mt-1 p-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500"
          data-testid="create-office-form-address-input"
          id="address-input-for-create-office-form"
          name="address"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group flex flex-col gap-y-2"
      >
        <label
          class="text-gray-700 font-medium"
          for="input-field"
        >
          Registry Phone Number
        </label>
        <input
          class="form-control mt-1 p-2 border border-gray-300 rounded-md focus:outline-none focus:ring-2 focus:ring-blue-500"
          data-testid="create-office-form-phone-number-input"
          id="registry-phone-number-input-for-create-office-form"
          name="registryPhoneNumber"
          type="text"
          value=""
        />
      </div>
      <div
        class="form-group flex flex-col gap-y-2"
      >
        <label
          class="flex items-center gap-2"
        >
          <input
            checked=""
            class="form-checkbox h-5 w-5 text-blue-600 border-gray-300 rounded focus:ring-blue-500"
            data-testid="create-office-form-is-active-checkbox"
            id="is-active-checkbox-input-for-create-office-form"
            name="isActive"
            type="checkbox"
          />
          <span
            class="text-gray-700 font-medium"
          >
            Is active
          </span>
        </label>
      </div>
      <div
        class="flex flex-row space-x-4"
      >
        <button
          class="px-4 py-2 bg-blue-500 text-white rounded disabled:bg-blue-900 disabled:text-gray-400 disabled:hover:cursor-not-allowed hover:bg-blue-600 focus:outline-none cursor-pointer w-full bg-blue-600 text-white hover:bg-blue-700"
          data-testid="submit-button"
          type="submit"
        >
          Submit
        </button>
        <button
          class="px-4 py-2 bg-blue-500 text-white rounded disabled:bg-blue-900 disabled:text-gray-400 disabled:hover:cursor-not-allowed hover:bg-blue-600 focus:outline-none cursor-pointer w-full bg-gray-600 text-white hover:bg-gray-700"
          data-testid="cancel-button"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</DocumentFragment>
`;
